common/beerocks/bcl/include/bcl/network/socket.h

```diff
@@ -43,28 +43,83 @@ typedef unsigned __int64 ssize_t;
-    Socket(const std::string &uds_path = std::string(), long readTimeout = 1000);
+    // explicit Socket(SOCKET s, long readTimeout = 1000);
+    // Socket(SOCKET s, const std::string &peer_ip, int port_port, long readTimeout = 1000);
+    // Socket(const std::string &uds_path = std::string(), long readTimeout = 1000);
+    static Socket *socketFactory(SOCKET s, long readTimeout = 1000);
     virtual ~Socket();
 
     enum SocketMode { SocketModeBlocking, SocketModeNonBlocking };
 
-    ssize_t getBytesReady();
-    ssize_t readBytes(uint8_t *buf, size_t buf_size, bool blocking, size_t buf_len = 0,
-                      bool isPeek = false, bool waitAll = false);
+    virtual ssize_t getBytesReady()                                      = 0;
+    virtual ssize_t readBytes(uint8_t *buf, size_t buf_size, bool blocking, size_t buf_len = 0,
+                              bool isPeek = false, bool waitAll = false) = 0;
     /**
      * @brief read and discard specified number of bytes
      *
      * @return error code if an error occured or number of discarded bytes otherwise
      */
     ssize_t discardBytes(size_t bytes_to_discard);
-    ssize_t writeBytes(const uint8_t *buf, size_t buf_len, int port = 0,
-                       struct sockaddr_in addr_in = {});
-    ssize_t writeString(std::string msg)
+    virtual ssize_t writeBytes(const uint8_t *buf, size_t buf_len, int port = 0,
+                               struct sockaddr_in addr_in = {}) = 0;
+    virtual ssize_t writeString(std::string msg)
     {
         return writeBytes((const uint8_t *)msg.c_str(), (size_t)msg.length());
     }
+    virtual bool setWriteTimeout(long msec) = 0;
+    virtual bool setReadTimeout(long msec)  = 0;
+    virtual void closeSocket()              = 0;
+    virtual bool isOpen()                   = 0;
+    virtual SOCKET getSocketFd()            = 0;
+    virtual std::string getError()          = 0;
+    virtual std::string getPeerIP()         = 0;
+    virtual int getPeerPort()               = 0;
+    virtual std::string getUdsPath()        = 0;
+    virtual void setPeerMac(const std::string &mac) = 0;
+    virtual std::string getPeerMac() =0;
+
+    virtual bool isAcceptedSocket() = 0;
+    virtual void setIsServer() = 0;
+
+#ifndef IS_WINDOWS
+    virtual size_t getBytesWritePending() = 0;
+#endif
+protected:
+    friend class SocketServer;
+    friend class SocketSelect;
+    virtual void setAcceptedSocket(bool) = 0;
+    virtual SOCKET getSocket() const = 0;
+    virtual void setUdsPath(const std::string &) = 0;
+/*
+    SOCKET m_socket = INVALID_SOCKET;
+    std::string m_error;
+    std::string m_peer_ip;
+    std::string m_uds_path;
+    int m_peer_port = 0;
+    std::string m_peer_mac;
+    bool m_accepted_socket  = false;
+    bool m_external_handler = false;
+    bool m_is_server        = false;
+
+private:
+    static int m_ref;
+*/
+};
+
+class DefaultSocket : public Socket {
```

Comment:
I think "DefaultSocket" is not a great name. I like "SocketImpl" similar to what @mariomaz uses in #1449 is more appropriate.
(nitpick)